sdk/ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala

```diff
@@ -394,13 +397,16 @@ class ContractsService(
 
           // TODO use `stream` when materializing DBContracts, so we could stream ActiveContracts
           val fv: Future[Vector[domain.ActiveContract[JsValue]]] =
-            unsafeRunAsync(searchDb_(fetch)(ctx, queryParams))
+            unsafeRunAsync(ctx.templateIds)(searchDb_(fetch)(ctx, queryParams))
 
           Source.future(fv).mapConcat(identity).map(\/.right)
         }
 
-        private[this] def unsafeRunAsync[A](cio: doobie.ConnectionIO[A]) =
+        private[this] def unsafeRunAsync[A](
+            templateIds: Set[TemplateId.RequiredPkg]
+        )(cio: doobie.ConnectionIO[A]) = lockSet.withLocksOn(templateIds) {
           dao.transact(cio).unsafeToFuture()
+        }
 
```

Comment:
I'm afraid that this locking strategy won't work as intended. `dao.transact` returns an `IO` computation object that is lazily evaluated, and `unsafeToFuture` merely spawns the computation inside the `Future` it returns. Yet, `withLocksOn` releases the locks as soon as the `Future` has been constructed, rather than when it completes.

You may want to specialize `LockSet` such that `block: => Future[A]` and then you release the locks only once the `Future` has completed (`.onComplete` or `.transform` are your friends.)